scipy/signal/spectral.py

```diff
@@ -281,19 +447,49 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
                       % (nperseg, axis, x.shape[axis], axis))
         nperseg = x.shape[-1]
 
+    # These checks only neccesary if x!=y
+    if not psd:
+        if y.shape[-1] < nperseg:
+            warnings.warn('nperseg = %d, is greater than y.shape[%d] = %d, '
+                          'using nperseg = y.shape[%d]'
+                          % (nperseg, axis, y.shape[axis], axis))
+            nperseg = y.shape[-1]
+
+        # Check if we can broadcast the remaining axes together
+        for a, b in zip(x.shape[-2::-1],y.shape[-2::-1]):
+            if a == 1 or b == 1 or a == b:
+                pass
+            else:
+                raise ValueError('x and y cannot be broadcast together.')
+
+        # Check if x and y are the same length
+        if x.shape[-1] != y.shape[-1]:
+            warnings.warn('Inputs x and y have different lengths along chosen '
+                          'axes.')
```

Comment:
There's a lot of warnings here. If ``x != y`` is not OK, just disallow it. If it's a valid use-case, it shouldn't raise warnings.